File: Task_2/firebase.js
const admin = require('firebase-admin');
var express = require('express');
var bodyParser = require('body-parser');
var app = express();

// Initialize Firebase
const serviceAccount = require('C:/Users/DiDa/Desktop/Task/pi-demo-5b792-firebase-adminsdk-t3d8v-669f4a9ef4.json');

admin.initializeApp({
  credential: admin.credential.cert(serviceAccount)
});

const db = admin.firestore();
const collectionRef = db.collection('Task_2');
app.use(bodyParser.json());
app.use(bodyParser.urlencoded({ extended: false }));

app.post("/firebase", () => {

    const observer = collectionRef.onSnapshot(querySnapshot => {
        console.log('Received collection snapshot');
        querySnapshot.forEach(docSnapshot => {
          console.log(`Document ID: ${docSnapshot.id}`);
          console.log(docSnapshot.data());
        });
      }, err => {
        console.log(`Encountered error: ${err}`);
      });
});

app.post("/stop", () => {
  console.log("Stopping server...");
  // Perform any necessary cleanup or finalization tasks here
  process.exit(0); // Exit the process with a success status code
});

app.listen(3000);
// Listen for data changes on the entire collection

File: Task_2/firebase.py
import firebase_admin
from firebase_admin import credentials
from firebase_admin import firestore
import requests
import json
# Initialize Firebase
cred = credentials.Certificate("C:/Users/DiDa/Desktop/Task/pi-demo-5b792-firebase-adminsdk-t3d8v-669f4a9ef4.json")
firebase_admin.initialize_app(cred)

db = firestore.client()

# Add Data
def add_data():
    doc_ref = db.collection('Task_2').document('Abdullah')
    data = {
        'name': 'Abdullah',
        'age':'22',
        'gender':'Male'
    }
    doc_ref.set(data)
    print("Data added.")
    return data

def initiate_post():
    res = requests.post('http://127.0.0.1:3000/firebase', json=add_data()) 
    return res.json()

def stop_server():
    res = requests.post('http://127.0.0.1:3000/stop')
    if res.status_code == 200:
        print("Server stopped successfully.")
    else:
        print("Failed to stop the server.")


if __name__ == '__main__':
    add_data()
    initiate_post()
    stop_server()